### PR TITLE
llvmPackages.bolt: init at 17.0.6

### LINKED ISSUE #176536: LLVM Bolt
**This project should be part of the `llvmPackage_14.bintools-unwrapped`, but it's missing.**

BOLT is a post-link optimizer developed to speed up large applications.

It achieves the improvements by optimizing application's code layout based on execution profile gathered by sampling profiler, such as Linux perf tool. An overview of the ideas implemented in BOLT along with a discussion of its potential and current results is available in [CGO'19 paper](https://research.fb.com/publications/bolt-a-practical-binary-optimizer-for-data-centers-and-beyond/).

**Metadata**

* homepage URL: https://github.com/llvm/llvm-project/tree/main/bolt
* source URL: https://github.com/llvm/llvm-project/tree/main/bolt
* license: Apache License v2.0 with LLVM Exceptions
* platforms: unix, linux, darwin.

### PR DESCRIPTION
**Pull request by pca006132**
## Description of changes

adds llvm-bolt for llvm 17.

Closes #176536. Note that we need quite a large patch, we probably want to upstream some changes to allow bolt standalone build.

Upstream PR: https://github.com/llvm/llvm-project/pull/87196

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pca006132** [comment]
I tested the binary a bit, llvm-bolt (with instrumentation) is working, but perf2bolt seems to be a bit problematic and I am not sure why. There is no error, but the profile result is not useful.

**Comment by Ericson2314** [comment]
I would create an upstream PR for that part of the patch and mention it here https://discourse.llvm.org/t/building-llvm-out-of-per-subproject-tar-balls-still-is-not-working/72829/11

I have helped shepherd such things upstream before.

**Comment by pca006132** [comment]
I think the patch should work for LLVM 16, but may need some minor tweaks. I can try to add it later today.

**Comment by alyssais** [comment]
It'd also be fine to have it be marked broken for LLVM 16.  The attribute just needs to exist for silly reasons.  mlir is a good example of how to have a common package definition between multiple LLVM versions, btw.

**Comment by pca006132** [comment]
Added the attribute for llvm 16 as well. It does not work because there is a large cmake file change in llvm 17, need large effort to make it work for llvm 16...

**Comment by pca006132** [comment]
btw not sure how should I name the commit(s), I can do a rebase later if needed.
while this adds bolt for llvm 16 as well, it is marked as broken so not sure if this counts as adding the package

File: pkgs/development/compilers/llvm/common/bolt/default.nix
Comment (alyssais) [inline pkgs/development/compilers/llvm/common/bolt/default.nix:39]:
```suggestion
    broken = lib.versionOlder version "17.0.0";
```

**Comment by alyssais** [comment]
I think the commit message should be titled "llvmPackages.bolt: init at 17.0.6", since there's no "llvm-bolt" attribute at the top level.  In the commit message body you could explain that there needs to be an llvmPackages_16.bolt attribute, even if it doesn't work, referencing the commit I referred to before.

BTW: what about the upstream PR @Ericson2314 suggested?

**Comment by RossComputerGuy** [comment]
Curious as to what the status of this is since I know someone who is wanting this in LLVM 18. If this is merged before 18 is merged then I can merge this into 18 and git.

**Comment by pca006132** [comment]
This is working, but perhaps people are hesitant to merge it due to the relatively large patch?

**Comment by RossComputerGuy** [comment]
Maybe, I'd consider this to be kinda small. I don't see anything bad and I'm testing this on an M1 Pro running NixOS. Only thing I can think of is adding `bolt_17` and `bolt_16` as aliases to the `bolt` packages in `llvmPackages_16/17`. I also didn't see it run any sort of tests, maybe getting it to run tests would be good.

**Comment by pca006132** [comment]
I thought of running tests, but it seems that bolt tests are disabled when not built together with clang. I can look into it later when I have more time.

**Comment by RossComputerGuy** [comment]
Alright then just add the aliases and I'd say this is all good.

**Comment by pca006132** [comment]
Do note that `bolt_16` is broken, so I think we should only add `bolt_17`?

**Comment by RossComputerGuy** [comment]
@pca006132 Yeah, add 17 first.

**Review by RossComputerGuy — APPROVED** [review]

**Comment by lovesegfault** [comment]
Why did we add a broken bolt for `llvmPackages_16`?

**Comment by alyssais** [comment]
> Curious as to what the status of this is

The status is, from my PoV, that we haven't had a response to our question about upstreaming the patch.

**Comment by pca006132** [comment]
@alyssais I think @Ericson2314 will upstream the patch?

**Comment by alyssais** [comment]
That wasn't my reading of his comment, but we can let him clarify.

**Comment by RossComputerGuy** [comment]
Please rebase and add bolt for LLVM 18 as well.

**Comment by pca006132** [comment]
It seems that the patch has to be different for different llvm versions, I think I will try to upstream this and see if they accept it.

**Comment by RossComputerGuy** [comment]
> I think I will try to upstream this and see if they accept it.

Please do, would love to have any patches we can get into upstream before any release.

**Comment by pca006132** [comment]
OK just added llvm 18, submitted a PR to upstream, tidied up the patch, as well as enabled RISCV support. The tablegen stuff is different for every LLVM version, so it is a bit annoying to keep them. Will see if upstream has a better way of doing this, maybe factor the cmake code.

**Comment by RossComputerGuy** [comment]
Yep, sounds good.

**Review by RossComputerGuy — APPROVED** [review]

**Comment by RossComputerGuy** [comment]
@Ericson2314 Yes, this PR was mentioned in the upstream PR.

**Comment by Ericson2314** [comment]
I mean do you have a link for here? Or else do I find the upstream PR?

**Comment by RossComputerGuy** [comment]
@pca006132 Yes, you can find it by scrolling up and seeing the line that says this PR was mentioned in another one.

**Review by alyssais — CHANGES_REQUESTED** [review]
Please link to the upstream PRs in comments where the patches are used, so it's easy to find them.

**Comment by pca006132** [comment]
@alyssais I will wait until the upstream give more comments about the patch and fix it all together.

File: pkgs/development/compilers/llvm/common/bolt/default.nix
Comment (RossComputerGuy) [inline pkgs/development/compilers/llvm/common/bolt/default.nix:13]: Please use the `src` derivation naming convention we use which is `"${pname}-src-${version}"`.

**Review by RossComputerGuy — APPROVED** [review]

**Comment by pwaller** [comment]
Ping, this has an approval, is this still blocked, if so, on what? Can it be merged as-is and then improved?

**Comment by pca006132** [comment]
@pwaller  Probably missing link from the upstream PR. I wanted to add a reproducer for the upstream PR using nix (because I somehow cannot make the docker build works), but I am stuck with the lit script. Was busy last month, and I am not at home right now, will revisit this next week.

**Comment by RossComputerGuy** [comment]
@pca006132 Is there any interest or progress in updating this PR?

**Comment by pca006132** [comment]
Sorry but probably no, I am busy with other projects.

I can close this one and let you open a new one?

**Comment by RossComputerGuy** [comment]
> I can close this one and let you open a new one?

Yeah, that sounds good. I already rebased this in my own fork to work with the most recent changes.

**Comment by pca006132** [comment]
Thanks a lot! And really sorry about having no progress for months.

**Comment by RossComputerGuy** [comment]
Yeah, it's all good. I've opened #326240.

**Pull request closed without merging**
